### ISSUE TITLE
Support for recursive structure

### ISSUE THREAD

**Issue by afsalthaj** [zio/zio-config#407]
As of now, there is no clear way to handle recursive config structures.

For example, the below example blows the stack.

```scala
package zio.config.typesafe

import zio.config._,  ConfigDescriptor._, zio.config.typesafe._

object RecursiveConfigSpec extends App with EitherSupport {
  case class Data(rows: Row)

  case class Row(
    id: Int,
    nested: Option[Data]
  )

  val row: ConfigDescriptor[Row] =  (int("id") |@| data.optional)(Row.apply, Row.unapply)
  def data: ConfigDescriptor[Data] = nested("rows")(row)(Data.apply, Data.unapply)

  val source = TypesafeConfigSource.fromHoconString(
    s"""
       |rows: {
       |  id : 1
       |  rows :{
       |    id : 2
       |  }
       |}
       |""".stripMargin
  )

read(data from source)
}

```

**Comment by vigoo** [comment]
I ran into the need for this when adding `zio-config` descriptors for `zio-aws` and looked into it a bit. I have a proposal with a proof of concept PR. Right now it only implements config description and reading, but if you like the approach (or we discuss a better one), I would be happy to finish it to cover everything.

I have introduced a new config descriptor combinator, `nestedRec` which is almost like `nested`, but:
- it is lazy evaluating it's inner descriptor so the specification does not get into an endless recursion
- in the `ReadModule`, `NestedRec` almost works like `Nested`, but it switches to _fail fast mode_ until the next `Zip`.

`Zip` in fail-fast mode does not evaluate it's `right` side if the `left` fails. When not using `nestedRec`, it works as before.

The idea is that an empty recursive nesting (`nestedRec("x")(nestedRec("x")...`) makes no sense and can never terminate as it never gets to a `Source` leaf node, so we assume that the recursion is always zipped together with some other non-recursive fields, and this is where we can detect if the inner level exists or not.

**Comment by afsalthaj** [comment]
We had to revert back some of the changes as part of RC28 in the process of fixing the bug (which I think was not looking good).

We will definitely follow up on the changes and the tests related to recursion as a later commit. The issue will be reopened.

Let's reopen the issue, and bring back recursion, if possible without reducing the error view (although we discussed this before, now we are changing our mind). 

I think its a good experimentation anyway. What I learned from this, we need zio-config to be aware of recursive behaviour, and we can add a recursion node that clearly indicates any error that happen after this node need to be short circuited. I am hoping users won't be so annoyed with `recurs(config)`.
I haven't tested but I also think the short circuit of error can be for those recursions that terminates with Optional node, and may keep working for those that terminates with, say, empty list.
